### ISSUE TITLE
Last element of GCS chunker can exceed chunk size

### ISSUE THREAD

**Issue by mpdn** [akka/alpakka#2267]
The GCS chunker emits the entire buffer when the input is closed: https://github.com/akka/alpakka/blob/master/google-cloud-storage/src/main/scala/akka/stream/alpakka/googlecloud/storage/impl/Chunker.scala#L52-L54

But this buffer might be larger than the chunk size if the chunker receives an individual byte string larger than the chunk size. This can eg. have the effect that GCS multipart uploads fail due to it creating too large HTTP entities for the final entity.

Instead, the chunker should ensure that the output chunks after the input has closed are smaller than the chunk size.

**Comment by seglo** [comment]
> Instead, the chunker should ensure that the output chunks after the input has closed are smaller than the chunk size.

Are you suggesting an exception be thrown if it's not smaller?  Perhaps we could break down the buffer into chunks and emit them all in this scenario instead.

**Comment by mpdn** [comment]
No, not an exception, just continuing breaking it into chunks as you suggest. I think changing the `emitChunk` method to something like this would work:

```
if (bufferBuilder.length > chunkSize) {
  val (chunk, nextBuffer) = bufferBuilder.result.splitAt(chunkSize)
  bufferBuilder.clear()
  bufferBuilder.append(nextBuffer)
  totalSize += chunk.size
  emit(out, Chunk(chunk))
} else if (!isClosed(in)) {
  pull(in)
} else {
  if (bufferBuilder.nonEmpty) {
    totalSize += bufferBuilder.length
    emit(out, Chunk(bufferBuilder.result(), Some(totalSize)))
    bufferBuilder.clear()
  }
  completeStage()
}
```

**Comment by ennru** [comment]
@mpdn Would you be in a position to suggest the necessary changes in a Pull Request?

**Comment by ennru** [comment]
Fixed with #2269
